src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -884,7 +884,24 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
         public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> source, int count)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            return SkipIterator<TSource>(source, count);
+
+            IList<TSource> sourceList = source as IList<TSource>;
+            return sourceList != null ? SkipList(sourceList, count) : SkipIterator<TSource>(source, count);
+        }
+
+        private static IEnumerable<TSource> SkipList<TSource>(IList<TSource> source, int count)
+        {
+            if (count < 0)
+            {
+                count = 0;
+            }
+
+            int sourceCount = source.Count;
```

Comment:
Yes, agree on this. Sideeffecting funcs are discouraged, but cannot be prevented.
When replacing iteration with indexing, we need to re-check Count to preserve existing behavior or we could potentially break some previously-working code.